Parquet writer tests for supported compression kinds

pdabre12 opened the pull request:
Testing different compressions kinds of native parquet writer.

pdabre12 commented:
@majetideepak  Please review, thanks !

pdabre12 commented:
With the recent changes in the `ThriftCodecToCompressionKind()` function in PageReader.h , an argument can no longer be passed into the `ThriftCodecToCompressionKind()` function, meaning that the compressionKind needs to be extracted in a different way.

pdabre12 commented:
One workaround could be adding the `ThriftCodecToCompressionKind()` function  as a helper function to the `ParquetTestBase.h` as done so here.

pdabre12 commented:
@czentgr Please take a look.

czentgr commented on velox/dwio/parquet/reader/ParquetReader.cpp:
```diff
@@ -785,6 +785,16 @@ size_t ParquetReader::numberOfRowGroups() const {
   return readerBase_->fileMetaData().row_groups.size();
 }
 
+thrift::CompressionCodec::type ParquetReader::getCompressionKind(
```
We didn't want to expose thrift structs/enums in the reader. Instead, it should be `common::CompressionKind` which is also returned for the DRWF reader. So the mapping from codec to kind has to occur here already.

pdabre12 replied:
> With the recent changes in the `ThriftCodecToCompressionKind()` function in PageReader.h , an argument can no longer be passed into the `ThriftCodecToCompressionKind()` function, meaning that the compressionKind needs to be extracted in a different way.

That was the plan all along , need to find out another way.

majetideepak replied:
Since there now is a re-use, let's move `common::CompressionKind ThriftCodecToCompressionKind` to ParquetReaderUtil.h/ParquetReaderUtil.cpp

czentgr replied:
Maybe I don't remember the conversation, but didn't we decide to not expose the `thrift::CompressionCodec::type` through the reader and instead return `common::CompressionKind`. The function would use a util function that does the codec conversion from ParquetReaderUtil.h/ParquetReaderUtil.cpp.

pdabre12 replied:
Yes , you are right, fixing that.

czentgr commented on velox/dwio/parquet/reader/ParquetReader.h:
```diff
@@ -126,6 +126,10 @@ class ParquetReader : public dwio::common::Reader {
   size_t numberOfRowGroups() const;
 
+  thrift::CompressionCodec::type getCompressionKind(
+      int32_t rowGroupIndex,
```
Do you want to use defaults = 0 here so that it matches the DWRF reader API when not providing groupIndex or colIndex?

majetideepak replied:
Parquet and Dwrf differ in terms of where the compression information is stored. For Dwrf, it is stored in the [Postscript](https://orc.apache.org/specification/ORCv1/) which is common for all the stripes.
For Parquet, the compression information is specific to the column in a RowGroup. In practice, all RowGroup columns use the same compression.
IMO we should not add a default here to ensure the client knows the compression information is tied to a column in a RowGroup.

majetideepak replied:
Let's go with Christian's suggestion for the ParquetReader API.
The ReaderBase API can have the arguments. See the example in the other comment.

pdabre12 commented:
Hello  @nmahadevuni, 
I am trying to test the parquet writer tests with the supported compression kinds. 
I am pulling in the compression codec used to write the file from the file metadata, my goal is to make sure that the file was written and read in the compression kind specified. 
For converting the compression Codec obtained from the file MetaData to the CompressionKind, I need to use the `thriftCodecToCompressionKind()` which is currently a member function of the PageReader class. 
Any suggestions on how I can tackle this without duplicating the `thriftCodecToCompressionKind()` helper function as is done in this PR? 

Thank you  for your time.

pdabre12 commented:
@majetideepak Can you please review this PR when you get the chance , need your inputs here.

yzhang1991 commented:
Hi @majetideepak, can you finish the review?

pdabre12 commented:
@yzhang1991 This PR (https://github.com/facebookincubator/velox/pull/7643) needs to be reviewed first.

majetideepak commented:
@pdabre12 can you please address the existing comments? Thanks.

pdabre12 commented:
> @pdabre12 can you please address the existing comments? Thanks.

Hi @majetideepak , I opened a new PR to address the comments. (https://github.com/facebookincubator/velox/pull/7643)

pdabre12 commented:
@majetideepak  @czentgr Please review again.

czentgr commented on velox/dwio/parquet/tests/writer/ParquetWriterTest.cpp:
```diff
@@ -0,0 +1,103 @@
+VELOX_INSTANTIATE_TEST_SUITE_P(
+    SupportedParquetWriterCompressionTestSuite,
+    SupportedParquetWriterCompressionTest,
+    testing::ValuesIn(SupportedParquetWriterCompressionTest::getTestParams()));
```
Make sure there is a new line at the end.

pdabre12 commented:
@czentgr Can you take another pass?

czentgr commented on velox/dwio/parquet/reader/ParquetReader.cpp:
```diff
@@ -803,6 +804,17 @@ size_t ParquetReader::numberOfRowGroups() const {
+                              .columns[colIndex]
+                              .meta_data.codec;
+  auto compressionKind = thriftCodecToCompressionKind(compressionCodec);
+  return compressionKind;
```
You can combine the previous and current line into
```
return thriftCodecToCompressionKind(compressionCodec);
```

majetideepak commented on velox/dwio/parquet/reader/ParquetReader.cpp:
```diff
@@ -803,6 +804,16 @@ size_t ParquetReader::numberOfRowGroups() const {
+common::CompressionKind ParquetReader::getCompressionKind(
+    int32_t rowGroupIndex,
+    int32_t colIndex) const {
+  auto compressionCodec = readerBase_->fileMetaData()
```
Let's follow the DWRF API as much as we can
```
  common::CompressionKind getCompression() const {
    return readerBase_->getCompressionKind(0, 0);
  }
```

majetideepak commented on velox/dwio/parquet/tests/writer/ParquetWriterTest.cpp:
```diff
@@ -0,0 +1,101 @@
+TEST_P(SupportedParquetWriterCompressionTest, compression) {
+  auto schema = ROW({"c0", "c1"}, {INTEGER(), DOUBLE()});
+  auto data = makeRowVector({
+      makeFlatVector<int32_t>(10, [](auto row) { return row + 5; }),
```
let's write `10'000` values. Make this a constant so that we can use it for the Reader as well.

majetideepak reviewed (commented):
@pdabre12 Looks good. Minor comments.

pdabre12 commented:
@majetideepak Addressed the comments, PTAL, thanks!

majetideepak commented on velox/dwio/parquet/tests/writer/ParquetWriterTest.cpp:
```diff
@@ -0,0 +1,101 @@
+TEST_P(SupportedParquetWriterCompressionTest, compression) {
+  auto schema = ROW({"c0", "c1"}, {INTEGER(), DOUBLE()});
+  const auto data = makeRowVector({
+      makeFlatVector<int32_t>(10'000, [](auto row) { return row + 5; }),
```
nit:  Use `const int64_t kRows = 10'000;`

majetideepak reviewed (approved):
@pdabre12 minor comment. Thanks!

pedroerp commented on velox/dwio/parquet/reader/ParquetReader.cpp:
```diff
@@ -803,6 +813,10 @@ size_t ParquetReader::numberOfRowGroups() const {
   return readerBase_->fileMetaData().row_groups.size();
 }
 
+common::CompressionKind ParquetReader::getCompressionKind() const {
```
feels to me like we should take the row id and column chunk id as a parameter here. 

Agreed with @majetideepak 's point that in most cases they happen to be the same, but in case they don't, the `pageReader->getCompressionKind()` will be pretty unintuitive and likely not do what the API client would expect.

majetideepak replied:
This is fair. I think in a follow-up PR, we should add API to get RowGroupMetaData, ColumnChunkMetaData and make `getCompressionKind()` part of `ColumnChunkMetaData` where it technically belongs.

pedroerp commented on velox/dwio/parquet/tests/writer/ParquetWriterTest.cpp:
```diff
@@ -0,0 +1,102 @@
+  CompressionKind compressionKind_;
+};
+
+TEST_P(SupportedParquetWriterCompressionTest, compression) {
```
to better test it, maybe we could encode each column with one compression type and have a single instantiation of this test?

majetideepak replied:
There is no column-level API available today since we use the Arrow bridge. The Writer Options API applies to all columns.
https://github.com/facebookincubator/velox/blob/main/velox/dwio/parquet/writer/Writer.h#L96

majetideepak replied:
@pdabre12 found a way to specify the column-level codec via the Arrow Options. We should be able to do this.

pedroerp commented on velox/dwio/parquet/tests/writer/ParquetWriterTest.cpp:
```diff
@@ -0,0 +1,102 @@
+      makeFlatVector<int32_t>(kRows, [](auto row) { return row + 5; }),
+      makeFlatVector<double>(kRows, [](auto row) { return row + 10; }),
+  });
+  auto filePath = fs::path(fmt::format("{}/test_close.txt", tempPath_->path));
```
could we use an in memory sink/source instead?

pdabre12 commented:
Addressed the comments, using an in-memory writer instead. 
Keeping the git log incase changes need to be rolled back. If changes are approved, will clean the git log. 

Please review, 
@majetideepak @pedroerp

pdabre12 commented:
@majetideepak Added column-level codec , PTAL. 
Thanks.

majetideepak commented on velox/dwio/parquet/tests/writer/ParquetWriterTest.cpp:
```diff
@@ -0,0 +1,112 @@
+  auto reader = createReaderInMemory(*sinkPtr, readerOptions);
+
+  ASSERT_EQ(reader->numberOfRows(), kRows);
+  auto type = reader->typeWithId();
```
ASSERT_EQ(type, schema);
We don't need to check per column.

majetideepak commented on velox/dwio/parquet/writer/Writer.cpp:
```diff
@@ -131,8 +131,15 @@ std::shared_ptr<WriterProperties> getArrowParquetWriterOptions(
   }
-  properties =
-      properties->compression(getArrowParquetCompression(options.compression));
+  if (options.columnCompressionsMap.empty()) {
```
I don't think it hurts to have the default compression.
Users could specify only a few columns.

majetideepak replied:
Can you verify what happens if only a few columns are specified?
Probably add 2 extra columns to the current test.

pdabre12 replied:
So I verified and it does not hurt to have the default compression. 
I added the 2 extra columns to the current test and both of them have a `compressionKind` of `common::CompressionKind_NONE` which is as desired.

I also specified both the `compression`  and the `columnCompressionsMap` writer options at the same time (only a few columns) and and the columns for which a compression wasn't specified in the `columnCompressionsMap`  matched the `compression` specified in the `compression writerOptions`. 

In short, it works.

majetideepak replied:
> I added the 2 extra columns to the current test and both of them have a compressionKind of common::CompressionKind_NONE `

Let's add a non-default to the current test for compression.

majetideepak commented on velox/dwio/parquet/tests/writer/ParquetWriterTest.cpp:
```diff
@@ -0,0 +1,110 @@
+  for (int i = 0; i < params.size(); i++) {
+    EXPECT_EQ(
+        reader->getCompressionKind(0, i),
+        (i < params.size()) ? params[i] : common::CompressionKind_NONE);
```
Can we use a non-default value for the compression?

pdabre12 replied:
Done, using `CompressionKind::CompressionKind_SNAPPY` as the compression.

majetideepak commented:
@pedroerp, can you take another look? Thanks.

majetideepak commented:
@pdabre12 Can you rebase with main? I added new API here https://github.com/facebookincubator/velox/pull/8486
to get the Column Compression. We should use that instead.

majetideepak commented on velox/dwio/parquet/tests/writer/ParquetWriterTest.cpp:
```diff
@@ -0,0 +1,112 @@
+
+  for (int i = 0; i < params.size(); i++) {
+    EXPECT_EQ(
+        reader->getCompressionKind(0, i),
```
`reader->fileMetaData().rowGroup(0).columnChunk(0).compression()`
